File: 卡码/代码随想录/1/11-图论/Go/dfs/LeetCode_0841.go
// 841. 钥匙和房间
package main

func canVisitAllRooms(rooms [][]int) bool {
	n := len(rooms)
	visited := make([]bool, n)
	for i := 0; i < n; i++ {
		visited[i] = false
	}
	var dfs func(key int)
	dfs = func(key int) {
		if visited[key] {
			return
		}
		visited[key] = true
		keys := rooms[key]
		for _, key := range keys {
			if !visited[key] {
				dfs(key)
			}
		}
	}
	dfs(0)
	for _, visit := range visited {
		if !visit {
			return false
		}
	}
	return true
}

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/图论/dfs/LeetCode_0841.java
package com.kurumi.leetCode.codeThoughts.图论.dfs;

import java.util.List;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 841. 钥匙和房间
 */
public class LeetCode_0841 {
    static class Solution {
        boolean[] visited;

        void dfs(List<List<Integer>> rooms, int key) {
            if (visited[key])
                return;
            visited[key] = true;
            List<Integer> keys = rooms.get(key);
            for (int k : keys) {
                if (!visited[k])
                    dfs(rooms, k);
            }
        }

        public boolean canVisitAllRooms(List<List<Integer>> rooms) {
            visited = new boolean[rooms.size()];
            dfs(rooms, 0);
            for (boolean visit : visited) {
                if (!visit)
                    return false;
            }
            return true;
        }
    }
}
